Add EIP: Multi-redeemable NFTs (#6672)
ethereum/EIPs: EIPS/eip-6672.md

---
eip: 6672
title: Multi-redeemable NFTs
description: An extension of ERC-721 which enables an NFT to be redeemed in multiple scenarios for either a physical or digital object
author: RE:DREAMER Lab <dev@redreamer.io>, Archie Chang (@ArchieR7) <archie@redreamer.io>, Kai Yu (@chihkaiyu) <kai@redreamer.io>, Yonathan Randyanto (@Randyanto) <randy@redreamer.io>
discussions-to: https://ethereum-magicians.org/t/eip-6672-multi-redeemable-nfts/13276
status: Draft
type: Standards Track
category: ERC
created: 2023-02-21
requires: 165, 721
---

## Abstract

> This EIP proposes an extension to the [ERC-721](./eip-721.md) standard for Non-Fungible Tokens (NFTs) to enable multi-redeemable NFTs. This extension would allow an NFT to be redeemed in multiple scenarios for either physical or digital objects and maintain a record of its redemption status on the blockchain.

Reviewer: I would include a short sentence defining what "redeem" means for this proposal.

Reviewer 2: added

## Motivation

[ERC-5560](./eip-5560.md) enables only one-time redemption of an NFT, which means the same NFT cannot be re-used for another redemption from different campaigns or events. Our proposed multi-redeemable NFT is an improved alternative to the [ERC-5560](./eip-5560.md) redeemable NFT.

> 

Reviewer: I'd like to see at least one example use case where you might want a multi-redeemable NFT.

Reviewer 2: added

## Specification

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119 and RFC 8174.

> **Every contract compliant with this EIP MUST implement `ERC-6672` and `ERC-721` interfaces.**

Reviewer: Because these are in backticks, you cane use the interface name directly:

```suggestion
**Every contract compliant with this EIP MUST implement `ERC6672` and `ERC721` interfaces.**
```

Reviewer 2: changed

    "type": "object",
    "properties": {
        "name": {
            "type": "string",
            "description": "Identifies the asset to which this NFT represents"
        },
        "description": {
            "type": "string",
            "description": "Describes the asset to which this NFT represents"
        },
        "image": {
            "type": "string",
            "description": "A URI pointing to a resource with mime type image/* representing the asset to which this NFT represents. Consider making any images at a width between 320 and 1080 pixels and aspect ratio between 1.91:1 and 4:5 inclusive."
        }
    },
    "redemptions": {
        "operator-tokenId-redemptionId": {
            "status": {
                "type": "string",
                "description": "The status of a redemption. Enum type can be used to represent the redemption status, such as redeemed, shipping, paid."
            },
            "description": {
                "type": "string",
                "description": "Describes the object that has been redeemed for an NFT, such as the name of an action figure series name or the color of the product."
            }
        }
    }
}
```

> 

Reviewer: I'd mention ERC-165 somewhere in the body, and include the interface ids.

Reviewer 2: for now we remove 165 from requirement first

## Rationale

### Redeem and Cancel Mechanism

> An operator SHALL only make an update to the redemption created by itself. Therefore, the `redeem()` and `cancel()` functions do not have an `_operator` parameter, and the `msg.sender` address MUST be used as the `_operator`.

Reviewer: If "An operator SHALL only make an update to the redemption created by itself." is a requirement, it should go in the Specification section. The rest is good Rationale though.

Reviewer 2: moved to Specification

- Operator wallet address, i.e. `_operator`

    It's possible that there are more than one party who would like to use the same NFT for redemption. For example, MisterPunks NFTs are eligible to be redemeed for both Event-X and Event-Y tickets, and each event's ticket redemption is handled by a different operator.

- Token identifier, i.e. `_tokenId`

    Each NFT holder will have different redemption records created by the same operator. Therefore, it's important to use token identifier as one of the keys.

- Redemption identifier, i.e. `_redemptionId`

    Using `_redemptionId` as one of the keys enables NFT holders to redeem the same NFT to the same operator in multiple campaigns. For example, Operator-X has 2 campaigns, i.e. campaign A and campaign B, and both campaigns allow for MisterPunks NFTs to be redemeed for physical action figures. Holder of MisterPunk #7 is eligible for redemption in both campaigns and each redemption is recorded with the same `_operator` and `_tokenId`, but with different `_redemptionId`.

### Metadata Choices (metadata extension)

> To provide additional information about a redemption, `redemptions` key-value pairs is added to the "[ERC-721](./eip-721.md) Metadata Extension". The key format for the redemptions key-value pairs MUST be standardized as `operator-tokenId-redemptionId`, where `operator` is the operator wallet address, `tokenId` is  the identifier of the token that has been redeemed, and `redemptionId` is the redemption identifier. The same combination of keys is used for the redemption flag, and the same rationale applies.

Reviewer: Seems like there's another requirement in here (the key-value format) that needs to be in Specification.

Reviewer 2: moved to Specification

The value of the key `operator-tokenId-redemptionId` is an object that contains the `status` and `description` of the redemption.

- Redemption status, i.e. `status`

    The redemption status can have a more granular level, rather than just being a flag with a `true` or `false` value. For instance, in cases of physical goods redemption, we may require the redemption status to be either `redeemed`, `paid`, or `shipping`. It is RECOMMENDED to use a string enum that is comprehensible by both the operator and the marketplace or any other parties that want to exhibit the status of the redemption.

- Description of the redemption, i.e. `description`

    The `description` can be used to provide more details about the redemption, such as information about the concert ticket, a detailed description of the action figures, and more.

## Backwards Compatibility

This standard is compatible with [ERC-721](./eip-721.md).

## Reference Implementation

> Here's an example of an [ERC-721](./eip-721.md) that includes the Multi-Redeemable ([IERC-6672](./eip-6672.md)) extension:

Reviewer: ```suggestion
Here's an example of an [ERC-721](./eip-721.md) that includes the Multi-Redeemable (`ERC6672`) extension:
```

Reviewer 2: changed

```solidity

> // SPDX-License-Identifier: UNLICENSED

Reviewer: If the code is inline in the proposal itself, it must be licensed under `CC0-1.0`. We're a bit more permissive for files under `assets/`, but it still needs to be an open source license (like apache or mit.) `UNLICENSED` literally means no license, so we can't distribute it.

Reviewer 2: set the license to `MIT` and put the contract under `assets/eip-6672/contracts/`

pragma solidity 0.8.16;

import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
import "./interfaces/IERC6672.sol";

> contract ERC6672 is ERC721, IERC6672 {

Reviewer: ```suggestion
contract ERC6672 is ERC721, ERC6672 {
```

Reviewer 2: I believe this is correct since ERC6672 is the implementation of the interface IERC6672

Reviewer: The interface name (as defined [here](https://github.com/ethereum/EIPs/pull/6672/files#diff-09ca8486e6261457e9870d748205103913ecfceabbca2c30c117c6c172f2ce03R41)) is `ERC6672`. Should probably keep that consistent in your reference implementation.

Reviewer 2: Aha nice catch, alrighty I have changed the interface contract code in the specification to `IERC`.
